### DISCUSSION TITLE
Coreg error (maybe)

### DISCUSSION THREAD

**Discussion by zeynabalshelh** [CoBrALab/RABIES#362] [General]
Hi,

I'm new to preclinical imaging and I'm trying to use this software to preprocess some rat scans. 

This is the command used. :

docker run -it --rm -v /home/zeynab/projects/pilot_usyd/data/sub-pilot1/:/sub-pilot1:ro -v /home/zeynab/projects/pilot_usyd/data/preprocess_outputs:/preprocess_outputs/ ghcr.io/cobralab/rabies:nightly preprocess /sub-pilot1 /preprocess_outputs --apply_STC --TR 2 --commonspace_reg masking=true,brain_extraction=false,template_registration=SyN,fast_commonspace=false

I am looking at the QC anat_inho_cor report and something clearly went wrong:

![sub-pilot1_T2w_inho_cor](https://github.com/CoBrALab/RABIES/assets/75798250/d8ae9907-f614-4b2f-89bc-c1d3674c67e0)

These are the other two reports generated. Where something is clearly not working correctly. 

![sub-pilot1_T2w_RAS_inho_cor_registration](https://github.com/CoBrALab/RABIES/assets/75798250/d13d980e-f1e1-4853-92ef-a980eabedd8b)
![_registration](https://github.com/CoBrALab/RABIES/assets/75798250/a09d73aa-0f0f-44e2-adde-1ae8373da6fa)

I'm not sure where to go from here. Please advise. 

Thanks!

**Comment by gdevenyi** [comment]
Looks like your input image is upside down. RABIES expects images to be in RAS orientation and follow the NIFTI standards. Perhaps whatever conversion tool you're using isn't standards compliant?

**Reply by Gab-D-G** [reply]
Actually its not that RABIES expect RAS orientation, because it automatically reorient to RAS **if a different orientation is detected**, using nibabel https://nipy.org/nibabel/reference/nibabel.funcs.html#as-closest-canonical. So the image orientation must be badly encoded.

**Reply by gdevenyi** [reply]
Right, incorrectly encoded according to the NIFTI standard is a better description.

**Reply by zeynabalshelh** [reply]
Thank you both! 

I used dcm2bids conversion tool. Do you recommend any particular conversion tool that is standards compliant?

**Reply by gdevenyi** [reply]
Your DICOMS are probably not standards compliant. Where did they come from.

**Reply by zeynabalshelh** [reply]
I see. They are collected using MR solutions 7T preclinical scanner.

**Reply by gdevenyi** [reply]
I have previously dealt with MR Solutions and had issues. I suggest you attempt to coordinate a discussion between their support staff and https://github.com/rordenlab/dcm2niix

**Reply by zeynabalshelh** [reply]
Do you know of any workarounds in the meantime? I tried to use fslreorient2std but I don't think that helped much:

![sub-pilot1_T2w_inho_cor](https://github.com/CoBrALab/RABIES/assets/75798250/fca0b84b-e4b7-4b6c-8776-e0f0debb97ef)

![sub-pilot1_T2w_RAS_inho_cor_registration](https://github.com/CoBrALab/RABIES/assets/75798250/81606a5b-587a-41f3-85c3-b16dc417cb60)

![_registration](https://github.com/CoBrALab/RABIES/assets/75798250/5620f817-73c0-4fc7-be3c-4be9e2977464)

**Reply by gdevenyi** [reply]
when loading the file in itksnap, it must be oriented according to the anatomical directions present in the viewer. I presume you can do that with some arcane FSL commands, but I don't know what. You may also be able to do the reorientation in itksnap.

**Reply by zeynabalshelh** [reply]
Hi everyone, I am working with MR solutions to fix this issue as we have some data collected using their machines and they are attempting to export the nii in the format that would be accepted by rabies. They have asked me to provide the exact nii format that is required by rabies. Can you please provide me with as much detail as possible? Thanks!

**Reply by Gab-D-G** [reply]
Hi, I would recommend consulting our example dataset for a reference https://zenodo.org/records/8349029. If the metadata in your image is not accurate, it is hard to tell what needs to be fixed without comparing properties with a reference. Ideally the image is in RAS orientation, meaning that the direction cosine should look like this:
-1 0 0
0 -1 0
0 0 1

But then in your case, the image axes don't correspond with the real brain axes. So you can compare after both images are in RAS orientation which axes are inaccurate compared to the reference I provided. That is all I can advise on my end, but still I am not sure this would provide the ideal fix to your issue, as you would need to manually swap axes.
I hope this helps!

**Reply by fengxiangmr** [reply]
Hi, 

I have compared the example dataset from the reference, the images are in RIP orientation.  Another T2w anatomical scan provided by Zeynab is LSP orientation, and SIGMA template is in LPI orientation. However all these datasets look the same layout in ITK-SNAP, where axial view is shown in the window of S-I R-L directions.

I have tried FSL FLIRT to work on our NII data (from MR Solutions scanner) and it worked, although the option of searchrx, searchry, searchrz needs to be taken care.

At MR Solutions platform, we are using the quadruped convention for the DICOM axis orientation for rodents, and biped convention for humans. 
we decided to use what’s right according to the DICOM standard:
https://dicom.nema.org/medical/dicom/current/output/chtml/part03/sect_C.7.6.2.html#sect_C.7.6.2

Here are q/a that explains the difference: 
https://biology.stackexchange.com/questions/54898/dorsal-vs-posterior-and-ventral-vs-anterior

It should be possible for the user to convert the image to whatever system they want.
For example, we could use this transformation matrix to convert the quadruped to biped orientation.
1 0 0 0
0 0 -1 0
0 1 0 0
0 0 0 1

Is the RABIES using ANTs for registration? is it possible to change some options/parameters in this step to make sure it's working on our data?

Thank you.

**Reply by gdevenyi** [reply]
Hi,

Please provide screenshots of the raw (post conversion) NIFTI data loaded in ITK-Snap, clearly showing the data and reference directions from ITK-Snap. If these directions are not correct, the fault lies in either the conversion software being used, or in the original raw data provided from the scanner.

Please also specify exactly how the data was collected and converted (orientation of the data, type of scanner, conversion softare/process).

**Reply by gdevenyi** [reply]
One followup, the reason FSL "works" is because the authors long ago added a feature during registration to try all possible orientations of coordinate space and choose the one that works, in order to work around the kind of broken orientation files I believe we're seeing here. RABIES (via ANTs) uses standard gradient descent techniques which assume the data is at least nominally the same orientation to start (i.e. you have validly oriented NIFTIs according to the standard)

**Reply by fengxiangmr** [reply]
below is the NIFTI data loaded in ITK-Snap, 

![image](https://github.com/CoBrALab/RABIES/assets/60684066/b765088d-552b-4a49-a9e2-3ae17c186be2)

and template data loaded, very similar views.  
![image](https://github.com/CoBrALab/RABIES/assets/60684066/73e69a51-6539-419b-9544-2d0663fdda91)

The NIFTI data is collected from MR Solutions 7T, using single-shot GE-EPI. 

The orientation is converted from the quadruped to biped, using the MRtrix3's command:
mrtransform 16598.nii.gz  -linear transform_quad_to_biped.txt  16598_reoriented.nii.gz

do you need also to consider the strides in the Nifti raw data?

Thank you.

**Reply by gdevenyi** [reply]
The orientation of the data appears correct.

I'm not sure what you mean by "strides" in this case. Can you please clarify

The collection of the entire body, rather than just the brain is likely to cause challenges with alignment. You can resolve this by manually (or semi manually) cropping the data to online include the brain.

Finally, we strongly recommend collecting isotropic data for structural alignment, as it greatly improves the ability to get good registration. We would rather a slightly lower resolution isotropic image than a higher resolution anisotropic image.

**Reply by zeynabalshelh** [reply]
Hi,

Using the nifti data that Xiang posted on top with the correct orientation does not work with the RABIES pipeline and we're not too sure why. The error log is attached. 

[rabies_preprocess.log](https://github.com/CoBrALab/RABIES/files/15124572/rabies_preprocess.log)

This is the command I am using. This command works with the Bruker data I have but it's not working with this MR solutions data. 

```sh
sudo docker run -it --rm -v /home/zeynab/projects/pilot_usyd/data/bids/:/bids_dataset:ro \
-v /home/zeynab/projects/pilot_usyd/data/preprocess_outputs/:/preprocess_outputs \
-v /usr/local/rat_sigma_atlas/:/rat_sigma_atlas:ro ghcr.io/cobralab/rabies:nightly \
-p MultiProc preprocess /bids_dataset /preprocess_outputs \
--anat_template=/rat_sigma_atlas/SIGMA-Rat-Brain-Templates-and-Atlases/SIGMA_Rat_Anatomical_Imaging/SIGMA_Rat_Anatomical_InVivo_Template/anatomy/SIGMA_InVivo_Anatomical_Brain_template.nii.gz \
--brain_mask=/rat_sigma_atlas/SIGMA-Rat-Brain-Templates-and-Atlases/SIGMA_Rat_Anatomical_Imaging/SIGMA_Rat_Anatomical_InVivo_Template/anatomy/SIGMA_InVivo_Anatomical_Brain_mask.nii.gz \
--WM_mask=/rat_sigma_atlas/SIGMA-Rat-Brain-Templates-and-Atlases/SIGMA_Rat_Anatomical_Imaging/SIGMA_Rat_Anatomical_InVivo_Template/anatomy/SIGMA_InVivo_Anatomical_Brain_wm.nii.gz \
--CSF_mask=/rat_sigma_atlas/SIGMA-Rat-Brain-Templates-and-Atlases/SIGMA_Rat_Anatomical_Imaging/SIGMA_Rat_Anatomical_InVivo_Template/anatomy/SIGMA_InVivo_Anatomical_Brain_csf.nii.gz \
--labels=/rat_sigma_atlas/SIGMA-Rat-Brain-Templates-and-Atlases/SIGMA_Rat_Brain_Atlases/SIGMA_Anatomical_Atlas/InVivo_Atlas/SIGMA_InVivo_Anatomical_Brain_Atlas.nii.gz \
--vascular_mask=/rat_sigma_atlas/SIGMA-Rat-Brain-Templates-and-Atlases/SIGMA_Rat_Anatomical_Imaging/SIGMA_Rat_Anatomical_InVivo_Template/anatomy/SIGMA_InVivo_Anatomical_Brain_mask.nii.gz
```

Is there anything you can see here that can tell us why it's not working?

**Reply by Gab-D-G** [reply]
Looks like neither nibabel nor SimpleITK can read the files.

nibabel:
```
nibabel.filebasedimages.ImageFileError: Cannot work out file type of "/bids_dataset/sub-stimpilot4/anat/sub-stimpilot4_T2w.nii.gz"
```

SimpleITK:
```
RuntimeError: Exception thrown in SimpleITK ImageFileReader_Execute: ../Code/IO/src/sitkImageReaderBase.cxx:97:
sitk::ERROR: The file "/bids_dataset/sub-stimpilot4/anat/sub-stimpilot4_T2w.nii.gz" does not exist.
```

Unfortunately that's all the log tell us, I can't know exactly why the files are not recognized. Perhaps you could share the header of the files? Any idea @gdevenyi ?

**Reply by gdevenyi** [reply]
That doesn't look to me like an issue with the file. This looks like either permissions or path issues.

Please check that you've mounted your paths correctly into the container.

Our recommended configuration is singularitry, not docker, can you please try that?

**Reply by zeynabalshelh** [reply]
Hi,

I changed the permissions of the bids folder and that seemed to work? I didn't realise it needed writing permission to the bids folder.

Ok, so we're running into another issue here:
![sub-stimpilot4_T2w_inho_cor](https://github.com/CoBrALab/RABIES/assets/75798250/5a3270ec-f3db-4afd-887d-5aeb9117cece)

I am using the same command as before but the correction is way off. 

Is this related to the orientation of the image?

**Reply by gdevenyi** [reply]
>  I didn't realise it needed writing permission to the bids folder.

Permissions with docker are complicated, its not really designed/intended for standard-user based use cases (again, singularity is the solution here)

As I said previously, your images are way over-sampled in terms of field-of-view, a basic assumption of registration is a (near) 1:1 relationship of the content of the images. If you look at the atlas FOV, it is just the brain, whereas you have the whole body and empty space around that. We can't supply a larger FOV atlas, so you will need to do some pre-cropping.

**Reply by Gab-D-G** [reply]
Hi @zeynabalshelh, first are you using the right template? (i.e. rat template if these are rat scans) If you are, then I would recommend trying the --anat_autobox parameter, which will crop images to remove extra non-brain tissue which is likely causing issues in your case.

**Reply by zeynabalshelh** [reply]
Thank you both!

Using --anat_autobox did not work but precropping the image did:
![sub-stimpilot4_T2w_inho_cor](https://github.com/CoBrALab/RABIES/assets/75798250/d7647778-5106-4244-b8f6-7f20f847da97)

Thanks for your help. We will try to implement your suggestion to use a smaller FOV and isotropic data for our future scans.

**Comment by zeynabalshelh** [comment]
Hi,

I am now testing this pipeline on data collected using a Bruker 9.4T. I converted the data using brkraw. I have tried the following:

```sh
sudo docker run -it --rm -v /home/zeynab/projects/UNSW/data/bids/:/bids_dataset:ro -v /home/zeynab/projects/UNSW/data/preprocess_outputs2/:/preprocess_outputs2 ghcr.io/cobralab/rabies:nightly -p MultiProc preprocess /bids_dataset /preprocess_outputs2 --anat_inho_cor method=Affine,otsu_thresh=4,multiotsu=false
```

 (i tried SYN, multiotsu on and off and differnt otsu thresholds).

For each one, it seems that the resampled mask is way off. 
![sub-Pilot1_ses-1_T2w_inho_cor](https://github.com/CoBrALab/RABIES/assets/75798250/56852198-e34c-4066-9a84-b6336c807804)

Each result from the different options I select are slightly different but they are all either masking a small area of the brain or sampling tissue outside the brain. I also tried using --anat_autobox. But I am having the same problem. 

I don't know how to fix this. Please help.

Thank you!

**Reply by Gab-D-G** [reply]
Is this a rat brain? if so you need to provide a rat atlas during preprocessing, because by default you will be using a mouse brain mask (hence the tiny mask you are obtaining). You can use the SIGMA rat atlas for instance https://zenodo.org/records/10635831. If you are using your own masks, can we see the template_files.png in the preprocess_QC_report?

**Comment by zeynabalshelh** [comment]
Hi,

Thank you for your response and patience with me as I learn these steps. Yes this is a rat brain. 
This is the template_files.png from the report:
![template_files](https://github.com/CoBrALab/RABIES/assets/75798250/52e4cd49-104b-46fa-9ab1-98bf6d5a4a18)

Is there something I am missing? Do I need to specify to use my own mask?

Thanks!

**Reply by Gab-D-G** [reply]
Hi,
That is the default mouse atlas within RABIES, not a rat atlas. So that is the origin of your error. You need to provide your own template files with the --anat_template parameters and associated mask parameters. See the CLI --help for preprocessing https://rabies.readthedocs.io/en/latest/running_the_software.html.
